File: app/src/main/java/com/github/overpass/gather/model/usecase/attachment/PhotosUseCase.java
package com.github.overpass.gather.model.usecase.attachment;

import android.content.ContentResolver;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;

import com.github.overpass.gather.model.commons.LiveDataUtils;
import com.github.overpass.gather.model.commons.exception.PhotoUploadException;
import com.github.overpass.gather.model.repo.meeting.MeetingRepo;
import com.github.overpass.gather.model.repo.upload.UploadImageRepo;
import com.github.overpass.gather.screen.auth.register.add.ImageUploadStatus;
import com.github.overpass.gather.screen.map.Meeting;
import com.github.overpass.gather.screen.meeting.chat.attachments.PhotoUploadStatus;

import java.util.UUID;

public class PhotosUseCase {

    private final MeetingRepo meetingRepo;
    private final UploadImageRepo uploadImageRepo;

    public PhotosUseCase(MeetingRepo meetingRepo,
                         UploadImageRepo uploadImageRepo) {
        this.meetingRepo = meetingRepo;
        this.uploadImageRepo = uploadImageRepo;
    }

    public LiveData<Meeting> getMeeting(String meetingId) {
        return meetingRepo.getLiveMeeting(meetingId);
    }


    public LiveData<PhotoUploadStatus> loadPhoto(ContentResolver contentResolver,
                                                 @NonNull Uri imageUri,
                                                 String meetingId) {
        return Transformations.switchMap(
                uploadImageRepo.saveImage(contentResolver, imageUri,
                        UploadImageRepo.BUCKET_MEETINGS, meetingId, UUID.randomUUID().toString()),
                status -> {
                    if (status instanceof ImageUploadStatus.Success) {
                        Uri uri = status.as(ImageUploadStatus.Success.class).getUri();
                        return Transformations.map(
                                meetingRepo.addPhoto(meetingId, uri.toString()),
                                photoAdded -> {
                                    if (photoAdded) {
                                        return new PhotoUploadStatus.Success();
                                    } else {
                                        return new PhotoUploadStatus.Error(new PhotoUploadException());
                                    }
                                }
                        );
                    } else if (status instanceof ImageUploadStatus.Error) {
                        Throwable throwable = status.as(ImageUploadStatus.Error.class).getThrowable();
                        return LiveDataUtils.just(new PhotoUploadStatus.Error(throwable));
                    } else {
                        return LiveDataUtils.just(new PhotoUploadStatus.Progress());
                    }
                }
        );
    }
}

File: app/src/main/java/com/github/overpass/gather/model/usecase/attachment/PhotosUseCase.kt
package com.github.overpass.gather.model.usecase.attachment

import androidx.lifecycle.LiveData
import com.github.overpass.gather.model.commons.Result
import com.github.overpass.gather.model.commons.SimpleResult
import com.github.overpass.gather.model.repo.meeting.MeetingRepo
import com.github.overpass.gather.model.repo.meeting.MeetingRepo2
import com.github.overpass.gather.model.repo.upload.UploadImageRepo
import com.github.overpass.gather.screen.map.Meeting
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import java.util.*
import javax.inject.Inject

class PhotosUseCase @Inject constructor(
        private val meetingRepo: MeetingRepo,
        private val meetingRepo2: MeetingRepo2,
        private val uploadImageRepo: UploadImageRepo
) {

    fun getMeeting(meetingId: String): LiveData<Meeting> {
        return meetingRepo.getLiveMeeting(meetingId)
    }

    @FlowPreview
    @ExperimentalCoroutinesApi
    suspend fun uploadPhoto(
            imageUri: ByteArray,
            meetingId: String
    ): Flow<Result<Unit>> = withContext(Dispatchers.IO) {
        val name = UUID.randomUUID().toString()
        uploadImageRepo.saveImage2(imageUri, UploadImageRepo.BUCKET_MEETINGS, meetingId, name)
                .map { result ->
                    when (result) {
                        is Result.Success ->
                            result.data
                                    .let { meetingRepo2.addPhotoUrl(meetingId, it) }
                                    .takeIf { it is SimpleResult.Error }
                                    ?.let { it as SimpleResult.Error }
                                    ?.run { Result.Error(exception) }
                                    ?: Result.Success(Unit)
                        is Result.Error -> Result.Error(result.exception)
                        is Result.Loading -> Result.Loading(result.percent)
                    }
                }
    }
}
